File: bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/immutable/Instrument.scala
package edu.gemini.model.p1.immutable

sealed abstract class Instrument(val site: Site, val id: String, val display: String, val gsa: String) extends Product with Serializable {
  def this(site: Site, id: String) = this(site, id, id, id)
  override def toString: String = display
}

import Site._

object Instrument {
  case object GmosNorth  extends Instrument(GN, "GMOS", "GMOS North", "GMOS-N")
  case object Gnirs      extends Instrument(GN, "GNIRS")
  case object Michelle   extends Instrument(GN, "Michelle", "Michelle", "michelle")
  case object Nifs       extends Instrument(GN, "NIFS")
  case object Niri       extends Instrument(GN, "NIRI")
  case object Dssi       extends Instrument(GN, "DSSI")
  case object Texes      extends Instrument(GN, "Texes", "Texes", "TEXES")
  case object Alopeke    extends Instrument(GN, "Alopeke", "'Alopeke", "Alopeke")
  case object Igrins2    extends Instrument(GN, "IGRINS-2", "IGRINS-2", "IGRINS-2")

  case object Flamingos2 extends Instrument(GS, "Flamingos2", "Flamingos2", "FLAMINGOS")
  case object Ghost      extends Instrument(GS, "GHOST")
  case object GmosSouth  extends Instrument(GS, "GMOSSouth", "GMOS South", "GMOS-S")
  case object Gpi        extends Instrument(GS, "GPI")
  case object Graces     extends Instrument(GN, "GRACES")
  case object Gsaoi      extends Instrument(GS, "GSAOI")
  case object Igrins     extends Instrument(GS, "IGRINS")
  case object Nici       extends Instrument(GS, "NICI")
  case object Phoenix    extends Instrument(GS, "Phoenix", "Phoenix", "PHOENIX")
  case object Trecs      extends Instrument(GS, "TReCS", "T-ReCS", "TReCS")
  case object Visitor    extends Instrument(GS, "Visitor")
  case object Zorro      extends Instrument(GS, "Zorro")
  case object MaroonX    extends Instrument(GN, "MAROON-X")

  def fromMutable(id: String): Instrument = id match {
    case "GMOS"       => GmosNorth
    case "GNIRS"      => Gnirs
    case "Michelle"   => Michelle
    case "NIFS"       => Nifs
    case "NIRI"       => Niri
    case "DSSI"       => Dssi
    case "Texes"      => Texes
    case "Alopeke"    => Alopeke
    case "Igrins2"    => Igrins2

    case "Flamingos2" => Flamingos2
    case "GHOST"      => Ghost
    case "GMOSSouth"  => GmosSouth
    case "GPI"        => Gpi
    case "GRACES"     => Graces
    case "GSAOI"      => Gsaoi
    case "IGRINS"     => Igrins
    case "NICI"       => Nici
    case "Phoenix"    => Phoenix
    case "TReCS"      => Trecs
    case "Visitor"    => Visitor
    case "Zorro"      => Zorro
    case "MAROON-X"   => MaroonX
    case _            => sys.error("Unsupported instrument")
  }

}

File: bundle/edu.gemini.pot/src/main/java/edu/gemini/pot/sp/Instrument.java
package edu.gemini.pot.sp;

import static edu.gemini.pot.sp.SPComponentType.*;
import edu.gemini.shared.util.immutable.ImOption;
import edu.gemini.shared.util.immutable.Option;
import edu.gemini.shared.util.immutable.None;
import edu.gemini.shared.util.immutable.Some;
import edu.gemini.spModel.core.Site;
import static edu.gemini.spModel.core.Site.*;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

/**
 * An enumeration of all instruments, which is a subset of all component types.
 */
public enum Instrument {
    AcquisitionCamera(INSTRUMENT_ACQCAM, GN, GS),
    Bhros(INSTRUMENT_BHROS),
    Flamingos2(INSTRUMENT_FLAMINGOS2, GS),
    Ghost(INSTRUMENT_GHOST, GS),
    GmosNorth(INSTRUMENT_GMOS, GN),
    GmosSouth(INSTRUMENT_GMOSSOUTH, GS),
    Gnirs(INSTRUMENT_GNIRS, GS),
    Gpi(INSTRUMENT_GPI, GS),
    Gsaoi(INSTRUMENT_GSAOI, GS),
    Michelle(INSTRUMENT_MICHELLE, GN),
    Nici(INSTRUMENT_NICI, GS),
    Nifs(INSTRUMENT_NIFS, GN),
    Niri(INSTRUMENT_NIRI, GN),
    Phoenix(INSTRUMENT_PHOENIX, GS),
    Texes(INSTRUMENT_TEXES),
    Trecs(INSTRUMENT_TRECS),
    Visitor(INSTRUMENT_VISITOR, GN, GS)
    ;

    public final SPComponentType componentType;
    public final Set<Site> sites;

    Instrument(SPComponentType type, Site... sites) {
        this.componentType = type;
        this.sites         = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(sites)));
    }

    public boolean existsAt(Site site) {
        return sites.contains(site);
    }

    /**
     * Wraps the instrument in an Option, to make it easy to pass to the
     * ISPFactory.createObservation methods.
     */
    public Option<Instrument> some() {
        return new Some<>(this);
    }

    /**
     * Convenience method for obtaining a none instance.
     */
    public static final Option<Instrument> none = None.instance();

    /**
     * Looks up the Instrument associated with an SPComponentType, if any.
     */
    public static Option<Instrument> fromComponentType(SPComponentType type) {
        return ImOption.fromOptional(
                   Arrays.stream(values())
                      .filter(i -> i.componentType == type)
                      .findFirst()
               );
    }
}
